### DISCUSSION TITLE
Comparing to Artemis

### DISCUSSION THREAD

**Discussion by baconcheese113** [heftapp/graphql_codegen#123] [General]
Hey! First off thanks for all your effort in both this lib and `graphql_flutter`....I'd probably go back to React Native without these. I read you were inspired by Relay and I think that's awesome! Relay is probably my favorite client-side graphql development tool, even more so than ApolloClient.

I switched from using `graphql_flutter` without static types to Artemis a month ago, and unfortunately I'd say it has just about as many cons as pros from my experience so far. Some of these are probably due to Dart limitations in the difficulty of working with json objects when compared to how easy typescript handles them....but still limitations regardless.

I made an Artemis [example project](https://github.com/baconcheese113/graphql-flutter-artemis-example#reasons-not-to-use-artemis) to document my findings and put some of the limitations there. I'm getting ready to make a similar example project with `graphql_codegen` and just wondering if the roadmap includes some features that Relay had and if the same limitations of Artemis exist in `graphql_codegen`.

[Here](https://github.com/baconcheese113/handle-it-app/blob/c2d524a56afd429fd0cc23a063ae3dbebbea2518/lib/feed/feed_home.dart#L19) and [here](https://github.com/baconcheese113/handle-it-app/blob/c2d524a56afd429fd0cc23a063ae3dbebbea2518/lib/feed/card/feed_card.dart#L23) are an example of how I was using graphql with `graphql_flutter`.

[Here](https://github.com/baconcheese113/madbirds-web/blob/master/client/src/components/scene/SceneEditor.tsx#L13) and [here](https://github.com/baconcheese113/madbirds-web/blob/master/client/src/components/scene/SceneViewer.tsx#L45) are an example of how I used graphql with ApolloClient. This is very similar to how I was using with Relay, except at the time I used the `fragmentContainer` HOC

With Artemis I'm constantly running into name collisions with sub fields defined in different fragments, and just using nested fragments in general. Also moving Queries/Fragments out of the file they're used in has made it harder to refactor code and prevent over-fetching once the variables are no longer used in that file. Here are some of the issues I had with Artemis https://github.com/comigor/artemis/issues/392 and https://github.com/comigor/artemis/issues/394 I'm at the point now where I'm spending 10 minutes adding new functionality and then 20 minutes wrestling with Artemis to generate types correctly for my nested fragments without name-collision errors, and since [I have to alias](https://github.com/baconcheese113/handle-it-app/blob/7125fc42e131ceee11dcf2e41e5dbc3e6b589247/lib/network/members/~graphql/members.fragments.graphql#L11) the sub-fields for them to work I'm sometimes fetching 5x more data for the same query without Artemis because each aliased field is resolved separately.

So my questions:

1) Do you have plans to allow Queries/Fragments to be defined in dart files? Can you elaborate about where the challenge is here and how it might be approached?
2) How do you avoid the field name collision problems that Artemis has?
3) Can you pass nested fragments down the widget hierarchy without issues?
4) Do you have plans to allow a dataId to automagically be added to types that contain it (similar to how __typename is), since it's required for cache normalization?

**Comment by budde377** [comment]
Thanks for reaching out, let me try and answer your questions:

> Do you have plans to allow Queries/Fragments to be defined in dart files? Can you elaborate about where the challenge is here and how it might be approached?

I've considered it, it amounts to analyzing the dart files and come up with a way to link the generated queries to the right methods/widgets. It is doable, but will be different than Relay because of the difference between the TS/Flow type system and Darts typesystem. Not unmanageable, but not started.

> How do you avoid the field name collision problems that Artemis has?

By a better implementation of how fragments are mapped to serializers. You can read more [here](https://budde377.medium.com/graphql-codegen-deep-dive-37eee522e4e5).

> Can you pass nested fragments down the widget hierarchy without issues?

I'd certainly hope so! A big part of the motivation for writing this library was that I disagreed with how Artemis handled fragments. If you experience any problems here, please open a issue.

> Do you have plans to allow a dataId to automagically be added to types that contain it (similar to how __typename is), since it's required for cache normalization?

Certainly sounds doable. Feel free to open an issue.

**Reply by baconcheese113** [reply]
> I've considered it, it amounts to analyzing the dart files and come up with a way to link the generated queries to the right methods/widgets. It is doable, but will be different than Relay because of the difference between the TS/Flow type system and Darts typesystem. Not unmanageable, but not started.

This is great to hear, I'm sure it'd be a huge selling point for `graphql_flutter` users considering making the switch to type generation!

And thanks for switching this to a discussion, this format is more appropriate. I'll open an issue re: dataIds

I'm working on that example project I mentioned right now, to decide whether to switch over my main project, and I plan to document the roadblocks I run into in this discussion if it remains open. I'll also be adding a similar `Reasons not to use` or `Pros/Cons` section in the README. On top of that I can try to document an Artemis -> graphql_codegen migration guide.

So far I really like the __typename generation and configuration option for output in a local `__generated__` folder, but I'm missing some of the aspects of Artemis like their [configuration options documentation](https://github.com/comigor/artemis/tree/master#configuration) and documentation on [how to put the schema in a custom location](https://github.com/comigor/artemis/tree/master#schema-mapping) it's a little trivial, but having the full list of the options with short descriptions and default values really helps users to orient around all the options

**Reply by budde377** [reply]
> I'm working on that example project I mentioned right now, to decide whether to switch over my main project, and I plan to document the roadblocks I run into in this discussion if it remains open. I'll also be adding a similar Reasons not to use or Pros/Cons section in the README. On top of that I can try to document an Artemis -> graphql_codegen migration guide.

Sounds amazing. I'm looking forward to hear what you're finding.


> So far I really like the __typename generation and configuration option for output in a local __generated__ folder, but I'm missing some of the aspects of Artemis like their [configuration options documentation](https://github.com/comigor/artemis/tree/master#configuration) and documentation on [how to put the schema in a custom location](https://github.com/comigor/artemis/tree/master#schema-mapping) it's a little trivial, but having the full list of the options with short descriptions and default values really helps users to orient around all the options


So for documentation, I'd love to improve this, but it's also an area where I think the community can really help. Any PRs improving docs are highly appreciated. Regarding schema mapping; the codegenerator doesn't care where you place your schemas and you can use scopes to  handle multiple [schemas](https://github.com/heftapp/graphql_codegen/tree/main/packages/graphql_codegen#multiple-schemas).

**Reply by baconcheese113** [reply]
I can definitely submit a PR to add a build.yaml options table, can you give me a hint where to look to find the existing options?

What if I want to place the schema.graphql in a graphql folder in the root dir? Is this possible?

**Reply by budde377** [reply]
Awesome! Here's the configuration [serialiser](packages/graphql_codegen_config/lib/config.dart) which contains all options. 

You can place it anywhere, you just need to update the `assetPath` option 😃

**Reply by baconcheese113** [reply]
Perfect, I'll work on that after this example project 💯 

Some problems I've run into (I can break these to issues if I'm not missing something):

1) [Unable to use hooks](https://github.com/baconcheese113/graphql_codegen_example/blob/master/lib/home/home_screen.dart#L17), get [useContext can only be called from the build method of HookWidget](https://stackoverflow.com/questions/64410479/usecontext-can-only-be-called-from-the-build-method-of-hookwidget) error
2) Builder stopped building my main query after adding another field to a fragment, it now only builds if I [add other fields](https://github.com/baconcheese113/graphql_codegen_example/blob/master/lib/home/~graphql/home_screen.query.graphql#L2) to the query alongside the fragment. I use this command to build `flutter pub run build_runner build --delete-conflicting-outputs` 

Other things I've noted:

1) The generated types implement all the fragments that are spread under them, so all the fields from the fragments are suggested and valid for the parent query. I get that this is how graphql works, but it also makes it more difficult to keep track of what fields are defined for which widgets. In relay the fields defined in a fragment are specific to that component, and further nested fragments are for child components. I think they do this through putting the generated fragment types on the prototype for the parent fragment/query, I wonder if there's a way to achieve this with Dart?

**Reply by budde377** [reply]
> [Unable to use hooks](https://github.com/baconcheese113/graphql_codegen_example/blob/master/lib/home/home_screen.dart#L17), get [useContext can only be called from the build method of HookWidget](https://stackoverflow.com/questions/64410479/usecontext-can-only-be-called-from-the-build-method-of-hookwidget) error

We're relying on the `flutter_hooks` package, so you'll need to extend the right widget, i.e. (`HookWidget`), https://pub.dev/packages/flutter_hooks.

> Builder stopped building my main query after adding another field to a fragment, it now only builds if I [add other fields](https://github.com/baconcheese113/graphql_codegen_example/blob/master/lib/home/~graphql/home_screen.query.graphql#L2) to the query alongside the fragment. I use this command to build flutter pub run build_runner build --delete-conflicting-outputs

I'm not sure I understand what you mean by "stopped building", can you elaborate? 

> The generated types implement all the fragments that are spread under them, so all the fields from the fragments are suggested and valid for the parent query. I get that this is how graphql works, but it also makes it more difficult to keep track of what fields are defined for which widgets. In relay the fields defined in a fragment are specific to that component, and further nested fragments are for child components. I think they do this through putting the generated fragment types on the prototype for the parent fragment/query, I wonder if there's a way to achieve this with Dart?


This would require us to create similar to a `useFragment` hook. I think it might be a good idea together with extracting the queries directly from Dart code, but ultimately it's a tradeoff between supporting `graphql` and `graphql_flutter`. So it is on my roadmap.

**Reply by baconcheese113** [reply]
Ah, didn't catch the setup needed to use Hooks, got it now and I'll include one hierarchy using hooks and one without in the example

I made issues for both the bug I'm experiencing with the builder https://github.com/heftapp/graphql_codegen/issues/127 and the feature request for more of a 1:1 relationship between queries/fragments and widgets https://github.com/heftapp/graphql_codegen/issues/128

**Edit** Sorry, just noticed we both created an issue for the same topic, feel free to close or edit however you prefer

**Reply by budde377** [reply]
#127 should now be fixed! Thanks for submitting good descriptive issues and reproducible examples.

**Reply by baconcheese113** [reply]
Noticed your new issues tag `graphql_switch`, what does that represent? 👀 

Wrapping up my example project, and I'm a little confused about the builder `client` options. Without `graphql` it seems like everything I need is built except for the Options classes. And without `graphql_flutter` none of the Query widgets are built, but the Options are. Is it realistic that users would use one without the other? Not sure how much overhead this is taking, can you elaborate a bit about that and the point of making it configurable?

**Reply by budde377** [reply]
> Noticed your new issues tag `graphql_switch`, what does that represent? 👀 


Becoming more relay like will require significant changes to the generated code and how other people are using it. To be as little disruptive to current workflows, I believe we can benefit from creating a new project with its own builder and possibly client abstraction. I've called this project for `graphql_switch`.

> Wrapping up my example project, and I'm a little confused about the builder `client` options. Without `graphql` it seems like everything I need is built except for the Options classes. And without `graphql_flutter` none of the Query widgets are built, but the Options are. Is it realistic that users would use one without the other? Not sure how much overhead this is taking, can you elaborate a bit about that and the point of making it configurable?

Many users are using `graphql` without `graphql_flutter` and so you can choose only to add the `graphql` client. This should not depend on the `graphql_flutter`. The flutter client does depend on the non flutter client and arguably we could just add that client when building for `graphql_flutter`.

**Reply by baconcheese113** [reply]
Finished the example app and write up https://github.com/baconcheese113/graphql_codegen_example I did attempt to switch the generated directory to an absolute one `/lib/__generated__` but was unable, the build finished successfully but didn't build any files

But overall, really enjoyed the experience with graphql_codegen so planning to migrate my main app over to using it next and hoping that it handles my nested fragments better than Artemis. I'll try to keep track of what changes I need to make and see if I can put together a migration guide.

Outside of types not including fields of nested fragments, I didn't come across other features that would make me recommend using Artemis over graphql_codegen

**Reply by budde377** [reply]
Thanks for documenting your journey! I'm looking forward to seeing the migration guide. 

Regarding the absolute paths, I looked at the example package and I noticed that it's using `v0.10.0-beta.7`. Can you try the following:

* Upgrading to the newly released `v0.10.0`
* Set the asset path, so if you're placing your graphql files in `/graphql` your asset path could be: `graphql/**`.

And try again.

**Reply by donmanueldev** [reply]
uff very good talk they had haha, I learned a lot about how graphql_codegen works, I've been using graphql_flutter with artemis for a while but I wanted to try this, it would be good to create more examples, I'm going to create examples using graphql-flutter and graphql_codegen, so that we have more documentation.

**Reply by baconcheese113** [reply]
The migration was actually very smooth, took a while because there were a lot of places that needed to be refactored, but it was also nice fixing a ton of hacks used to get Artemis working.

Migration guide is [here](https://github.com/baconcheese113/graphql_codegen_example#migration-guide-from-artemis-to-graphql_codegen)

**Reply by budde377** [reply]
Great work!

**Comment by vasilich6107** [comment]
Small comment about Artemis.
The main issue of all is a lack of maintainers and contributors)
We are welcoming all PRs but there is a lot of them coming into pipeline. So current state is “it somehow works”)
There is definitely a room for improvements but the cost of improvements in comparison to profit is pretty high.

I do not have time to make global shifts in Artemis library and I’m ok with how it works now)
I’m trying to fix critical bugs in Artemis. That’s all I have time for :-)

**Reply by baconcheese113** [reply]
Totally understand, and you're not even the repository owner so I imagine there are some other collaboration challenges! The reason for this discussion was mostly to talk about the current state of graphql type generators for Dart and see which aligned more with the functionality I needed

For new Dart users (myself included) it can be difficult to decide which library to choose for a new project, so I'm hoping this discussion and my example projects with both Artemis and graphql_codegen help users quickly understand the limitations of both and decide which is better suited to their project

@budde377 was an Artemis user and was inspired by it when creating graphql_codegen, so we're very thankful for your efforts (along with the efforts of the other contributors). You might find #142 interesting as well

**Reply by oyatrides** [reply]
> For new Dart users (myself included) it can be difficult to decide which library to choose for a new project, so I'm hoping this discussion and my example projects with both Artemis and graphql_codegen help users quickly understand the limitations of both and decide which is better suited to their project

It is exactly my case, and this discussion tremendously helped me (and soothed my mind) to validate my choice.
Thank you a lot for your work.
